File: src/application/usecase/role/update.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::RoleRepository;
use crate::domain;

pub struct UpdateRoleInput {
    pub name: String,
    pub description: String,
}

pub struct UpdateRoleOutput {
    pub role: domain::Role,
}

#[async_trait]
pub trait UpdateRoleUseCase: Send + Sync {
    async fn execute(&self, id: &str, input: UpdateRoleInput) -> Result<UpdateRoleOutput, AppError>;
}

pub struct UpdateRoleUseCaseImpl {
    role_repo: Arc<dyn RoleRepository>,
}

impl UpdateRoleUseCaseImpl {
    pub fn new(role_repo: Arc<dyn RoleRepository>) -> Self {
        UpdateRoleUseCaseImpl { role_repo }
    }
}

#[async_trait]
impl UpdateRoleUseCase for UpdateRoleUseCaseImpl {
    async fn execute(&self, id: &str, input: UpdateRoleInput) -> Result<UpdateRoleOutput, AppError> {
        let mut role = self.role_repo.find_by_id(id).await?;

        if input.name != role.name {
            let existing = self.role_repo.find_by_name(&input.name).await;
            if existing.is_ok() {
                return Err(AppError::from(domain::Error::AlreadyExists));
            }
        }

        role.update_info(input.name, input.description)?;
        self.role_repo.update(&role).await?;
        Ok(UpdateRoleOutput { role })
    }
}

File: internal/application/usecase/role/update.go
package role

import (
	"context"

	"go-clean-template/internal/domain"
)

type UpdateRoleInput struct {
	Name        string
	Description string
}

type UpdateRoleOutput struct {
	Role *domain.Role
}

type UpdateRoleUseCase interface {
	Execute(ctx context.Context, id string, input *UpdateRoleInput) (*UpdateRoleOutput, error)
}

type updateRoleUseCase struct {
	roleRepo domain.RoleRepository
}

func NewUpdateRoleUseCase(roleRepo domain.RoleRepository) UpdateRoleUseCase {
	return &updateRoleUseCase{roleRepo: roleRepo}
}

func (uc *updateRoleUseCase) Execute(ctx context.Context, id string, input *UpdateRoleInput) (*UpdateRoleOutput, error) {
	role, err := uc.roleRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if input.Name != role.Name {
		existing, _ := uc.roleRepo.FindByName(ctx, input.Name)
		if existing != nil {
			return nil, domain.ErrAlreadyExists
		}
	}
	if err := role.UpdateInfo(input.Name, input.Description); err != nil {
		return nil, err
	}
	if err := uc.roleRepo.Update(ctx, role); err != nil {
		return nil, err
	}
	return &UpdateRoleOutput{Role: role}, nil
}
